src/RoomMembers.ts

```diff
@@ -0,0 +1,270 @@
+    private perRoom: Map<string /* room id */, RoomMembers> = new Map();
+    private readonly cbHandleEvent;
+    constructor(private client: MatrixClient) {
+        // Listen for join events.
+        this.cbHandleEvent = this.handleEvent.bind(this);
+        client.on("room.event", this.cbHandleEvent);
+    }
+
+    /**
+     * Start listening to join/leave events in a room.
+     */
+    public addRoom(roomId: string) {
+        if (this.perRoom.has(roomId)) {
+            // Nothing to do.
+            return;
+        }
+        this.perRoom.set(roomId, new RoomMembers());
+    }
+
+    /**
+     * Stop listening to join/leave events in a room.
+     *
+     * Cleanup any remaining data on join/leave events.
+     */
+    public removeRoom(roomId: string) {
+        this.perRoom.delete(roomId);
+    }
+
+    public cleanup(roomId: string) {
+        this.perRoom.get(roomId)?.cleanup();
+    }
+
+    /**
+     * Dispose of this object.
+     */
+    public dispose() {
+        this.client.off("room.event", this.cbHandleEvent);
+    }
+
+    /**
+     * Return the date at which user `userId` has joined room `roomId`, or `null` if
+     * that user has joined the room before Mjölnir started watching it.
+     *
+     * @param roomId The id of the room we're interested in.
+     * @param userId The id of the user we're interested in.
+     * @returns a Date if Mjölnir has witnessed the user joining the room,
+     * `null` otherwise. The latter may happen either if the user has joined
+     * the room before Mjölnir or if the user is not currently in the room.
+     */
+    public getUserJoin(user: { roomId: string, userId: string }): Date | null {
+        const { roomId, userId } = user;
+        const ts = this.perRoom.get(roomId)?.get(userId) || null;
+        if (!ts) {
+            return null;
+        }
+        return new Date(ts);
+    }
+
+    /**
+     * Get the users in a room, ranked by most recently joined to oldest join.
+     *
+     * Only the users who have joined since the start of Mjölnir are returned.
+     */
+    public getUsersInRoom(roomId: string, since: Date, max = 100): Join[] {
+        const inRoom = this.perRoom.get(roomId);
+        if (!inRoom) {
+            return [];
+        }
+        return inRoom.members(since, max);
+    }
+
+    /**
+     * Record join/leave events.
+     */
+    public async handleEvent(roomId: string, event: any, now?: Date) {
+        if (event['type'] !== 'm.room.member') {
+            // Not a join/leave event.
+            return;
+        }
+
+        const members = this.perRoom.get(roomId);
+        if (!members) {
+            // Not a room we are watching.
+            return;
+        }
+        const userId = event['state_key'];
+        if (!userId) {
+            // Ill-formed event.
+            return;
+        }
+
+        const userState = event['content']['membership'];
+        let prevMembership = event['unsigned']?.['prev_content']?.['membership'] || "leave";
```

Comment:
shortcoming of my matrix knowledge; why is previous state important?

Reply:
A user changing display name will re-trigger a `join`, if my memory serves. A user may also be banned after leaving, but this doesn't change the fact that it has already left. There may be other non-trivial transitions.

See https://spec.matrix.org/v1.2/client-server-api/#mroommember .